### PR TITLE
2282-file-upload-integration-test

### LINKED ISSUE #2282: As technical lead, I want a file upload integration test
**Description:**
_Provide a brief background and justification for this issue_
#2141 spin-off, we want a full end-to-end test flow for the critical data analyst user path.

**Acceptance Criteria:** 

- [ ] An authenticated, approved test user can successfully upload a file.
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ]  Implement mechanism for searching year/quarter prep for file-upload
- [ ]  Implement test(s) for normal-case file upload with standard test file
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**

**Supporting Documentation:**

**Open Questions:**

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #2282 _

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] [**_insert ACs here_**]
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Comment by raftmsohani** [comment]
Created the logic, trying to build the local env for getting logs.

File: tdrs-frontend/cypress/e2e/accounts/accounts.feature [hunk 1]
```diff
@@ -39,3 +39,15 @@ Feature: Users can create and manage their accounts
         Then 'new-cypress@teamraft.com' can see the hompage
         When The admin sets the approval status of 'new-cypress@teamraft.com' to 'Deactivated'
         Then 'new-cypress@teamraft.com' cannot log in
+    Scenario: A user can upload a file
```
Comment (jtimpe) [inline tdrs-frontend/cypress/e2e/accounts/accounts.feature]: can we move this scenario to it's own feature file, in it's own folder? and move the steps to their own file as well

File: tdrs-frontend/cypress/e2e/accounts/accounts.feature [hunk 2]
```diff
@@ -39,3 +39,15 @@ Feature: Users can create and manage their accounts
...
+        Given The admin logs in
+        And 'new-cypress@teamraft.com' is in begin state
+        When 'new-cypress@teamraft.com' visits the home page
+        And 'new-cypress@teamraft.com' logs in
+        Then 'new-cypress@teamraft.com' requests access
+        And The admin sets the approval status of 'new-cypress@teamraft.com' to 'Approved'
```
Comment (jtimpe) [inline tdrs-frontend/cypress/e2e/accounts/accounts.feature]: is there a way we can skip the `is in a begin state` through `admin sets the status to 'Approved'` by having a user `is in an approved state` step?

File: tdrs-frontend/cypress/e2e/accounts/accounts.js [hunk 3]
```diff
@@ -57,3 +57,32 @@ Then('{string} sees the request still submitted', (username) => {
...
+Then('{string} can browse upload file form', (username) => {
+  cy.get('#reportingYears').should('exist').select('2023')
+  cy.get('#quarter').should('exist').select('Q1')
+  cy.get('button').contains('Search').should('exist')
+})
+
+When('{string} uploads a file', (username) => {
+  cy.get('button').contains('Search').should('exist').click()
+  cy.get('#closed-case-data').selectFile('../tdrs-backend/tdpservice/parsers/test/data/small_correct_file',{ action: 'drag-drop' })
+  cy.get('button').contains('Submit Data Files').should('exist').click()
```
Comment (jtimpe) [inline tdrs-frontend/cypress/e2e/accounts/accounts.js]: i really like the use of assertions throughout the `When` steps!

**Review by jtimpe — APPROVED** [review]
Well done! I had a couple small suggestions regarding test organization but otherwise looks good and works well

File: tdrs-frontend/cypress/e2e/accounts/accounts.js [hunk 4]
```diff
@@ -57,3 +57,32 @@ Then('{string} sees the request still submitted', (username) => {
...
+
+Then('{string} can browse upload file form', (username) => {
+  cy.get('#reportingYears').should('exist').select('2023')
+  cy.get('#quarter').should('exist').select('Q1')
+  cy.get('button').contains('Search').should('exist')
+})
+
+When('{string} uploads a file', (username) => {
+  cy.get('button').contains('Search').should('exist').click()
+  cy.get('#closed-case-data').selectFile('../tdrs-backend/tdpservice/parsers/test/data/small_correct_file',{ action: 'drag-drop' })
```
Comment (elipe17) [inline tdrs-frontend/cypress/e2e/accounts/accounts.js]: Can we submit the data file to its appropriate section so we can verify the records are created?

File: tdrs-frontend/cypress/e2e/accounts/accounts.feature (hunk 2 above)
Comment (raftmsohani) [reply tdrs-frontend/cypress/e2e/accounts/accounts.feature]: should be able to do that

File: tdrs-frontend/cypress/e2e/accounts/accounts.feature (hunk 1 above)
Comment (raftmsohani) [reply tdrs-frontend/cypress/e2e/accounts/accounts.feature]: Of course, I like that

File: tdrs-frontend/cypress/e2e/accounts/accounts.js (hunk 4 above)
Comment (raftmsohani) [reply tdrs-frontend/cypress/e2e/accounts/accounts.js]: Yes it is possible, however I will consider this PR to be base work which can be extended later. I suggest we create another ticket to submit a specific section file and confirm the correct result.
One point though: we should not overkill and submit multiple files from different sections. So my suggestion is to submit a file that is taking longer time, and limit when that test is running only to when backend::parsing logic is changed, then we might even be able to have a performance test and measure how long it is taking for the file to be parsed?

**Review by andrew-jameson — APPROVED** [review]

**Pull request merged**